### PR TITLE
Refactor DamagedByTerrain and introduce D2kBuilding

### PR DESCRIPTION
**Pull request by abcdefg30**
Otherwise you can keep buildings at full health with just one cell of concrete:
grafik
The building starts out damaged, but once repaired it will stay at full health.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 1]
```diff
@@ -92,8 +92,9 @@ void ITick.Tick(Actor self)
 			if (!self.IsInWorld)
 				return;
 
-			var t = self.World.Map.GetTerrainInfo(self.Location);
-			if (!Info.Terrain.Contains(t.Type))
+			var map = self.World.Map;
+			var cells = self.OccupiesSpace.OccupiedCells();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: At the risk of scope creeping things:
* We should try to cache this calculation so we only evaluate it if the actor moves
* We should apply the same fix to`GrantConditionOnTerrain`.

Would it be worth introducing a new `INotifyOccupiedCellsChanged` to simplify these?
This is a distinct usecase from `INotifyMoving`, etc.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Caching was my initial idea as well, but the terrain can change underneath while not moving. Would have to look into somehow hooking into `CellEntryChanged`.

**Comment by abcdefg30** [comment]
Rebased & updated.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 2]
```diff
@@ -32,68 +32,35 @@ class DamagedByTerrainInfo : ConditionalTraitInfo, Requires<IHealthInfo>, Requir
...
+			if (IsTraitDisabled || --damageTicks > 0)
 				return;
 
 			// Prevents harming cargo.
 			if (!self.IsInWorld)
 				return;
 
-			var t = self.World.Map.GetTerrainInfo(self.Location);
-			if (!Info.Terrain.Contains(t.Type))
+			if (cachedLocation != self.Location)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: This will regress the case where tiberium grows underneath where infantry are standing. For the sake of getting this PR merged without further delay IMO we should just remove the caching.

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		public readonly BitSet<DamageType> DamageTypes = default(BitSet<DamageType>);
+
+		[Desc("Terrain types where the actor will take damage.")]
+		public readonly string[] DamageTerrainTypes = { "Rock" };
+
+		[Desc("Percentage health below which the actor will not receive further damage.")]
+		public readonly int DamageThreshold = 50;
+
+		[Desc("Inflict damage down to the DamageThreshold when the actor gets created on damaging terrain.")]
+		public readonly bool StartOnThreshold = true;
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:37]: AFAIK this isn't used / useful for the D2k building usecase, so could perhaps be removed?

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs [hunk 4]
```diff
@@ -57,12 +67,51 @@ public D2kBuilding(ActorInitializer init, D2kBuildingInfo info)
...
+				if (template.PickAny)
+				{
+					// Fill the footprint with random variants
+					foreach (var c in info.Tiles(self.Location))
+					{
+						if (!map.Contains(c) || map.CustomTerrain[c] != byte.MaxValue)
+							continue;
+
+						// Don't place under other buildings
+						if (self.World.ActorMap.GetActorsAt(c).Any(a => a != self && a.TraitOrDefault<Building>() != null))
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:91]: Depending on what we do with #18824, this could use `self.World.ActorMapIsBuildingBlocked` (after moving this block above the `base.AddedToWorld`)?

**Comment by pchote** [comment]
(Label is just for the first point, the other two are non-blocking questions/discussion points)

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:37]: Yeah, I wondered about this too, but then thought this might be useful in the future if we want a building to start with full health but then take damage. I can remove it as practically dead code if you want.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs (hunk 2 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Ah right, that is why I wanted to look into `CellLayer.CellEntryChanged`. I think the issue here was that there are several layers (can I cover all?) and just using the one for resources might be too specific. (Even though it would work for our cases?)

Comment (pchote) [reply OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: All the layers should end up modifying the same `Map.CustomTerrain`. We can have the `ActorMap` add a listener to that and then notify any actors on the changed cells. Straightforward in theory, but I suspect this will end up being big enough that we won't want to tackle it right now.

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:37]: Ok, lets keep it then.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs (hunk 2 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Sounds good for a follow-up. Removed for now.

**Review by pchote — APPROVED** [review]

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+					{
+						var c = self.Location + new CVec(i % template.Size.X, i / template.Size.X);
+						if (!map.Contains(c) || map.CustomTerrain[c] != byte.MaxValue)
+							continue;
+
+						// Don't place under other buildings
+						if (self.World.ActorMap.GetActorsAt(c).Any(a => a != self && a.TraitOrDefault<Building>() != null))
+							continue;
+
+						layer.AddTile(c, new TerrainTile(template.Id, (byte)i));
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:110]: I think `layer` can be null in here.

File: mods/d2k/rules/arrakis.yaml [hunk 6]
```diff
@@ -134,17 +134,17 @@ sietch:
 	Inherits: ^Building
 	Tooltip:
 		Name: Fremen Sietch
-	Building:
+	D2kBuilding:
 		Footprint: xx xx
 		Dimensions: 2,2
 		TerrainTypes: Cliff
+		DamageTerrainTypes:
```
Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml:141]: Not sure if this is necessary. The tiles where it fits are all just cliff anyway.

File: OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs (hunk 5 above)
Comment (Mailaender) [reply OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs:110]: Actually, I am wrong. Nevermind.

**Review by Mailaender — APPROVED** [review]
Works as promised.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f0ccca803a4125d4e0d3ca85ba571c4f6bd5c62a...ed9151d2acaf6fdacb2273198b447d846b3509f8)

File: mods/d2k/rules/arrakis.yaml (hunk 6 above)
Comment (abcdefg30) [reply mods/d2k/rules/arrakis.yaml:141]: I had to place the sietch in the fourth atreides mission on passable ground so that the AI would attack it. Hence it would take damage in that mission.
